sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -650,7 +651,7 @@ class CodegenContext extends Logging {
     case FloatType =>
       val clsName = SQLOrderingUtil.getClass.getName.stripSuffix("$")
       s"$clsName.compareFloats($c1, $c2)"
-    // use c1 - c2 may overflow
+    case st: StringType => s"$c1.semanticCompare($c2, ${st.collationId})"
```

Comment:
shall we still generate java code `$c1.equalsTo($c)` if the string collation is UCS_BASIC?

Reply:
Done. Did the same thing for both compare and equals.